File: apps/mobile/ios/OneKeyWallet/BackgroundRunnerModule.m
#import "BackgroundRunnerModule.h"
#import "BackgroundReactNativeDelegate.h"

#import <Expo/Expo-Swift.h>
#import <ExpoModulesCore/ExpoModulesCore-Swift.h>
#import <ReactAppDependencyProvider/RCTAppDependencyProvider.h>
#import <React/RCTEventEmitter.h>

@implementation BackgroundRunnerModule {
    BOOL _hasListeners;
}

static RCTRootView *backgroundRootView = nil;

RCT_EXPORT_MODULE();

+ (instancetype)sharedInstance {
    static BackgroundRunnerModule *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"toUI", @"toBackground"];
}

- (void)startBackgroundRunner {
    // Prevent multiple inits
    if (backgroundRootView != nil) {
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        // Create BackgroundReactNativeDelegate
        BackgroundReactNativeDelegate *delegate = [[BackgroundReactNativeDelegate alloc] init];
        ExpoReactNativeFactory *factory = [[ExpoReactNativeFactory alloc] initWithDelegate:delegate];
        [delegate setDependencyProvider: [[RCTAppDependencyProvider alloc] init]];
        bindReactNativeFactory(factory);

        UIWindow *window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
        // This assumes JS entry point is "background"
        [factory startReactNativeWithModuleName:BACKGROUND_BUNDLE_NAME inWindow:window launchOptions:nil];
    });
}

RCT_EXPORT_METHOD(sendToUI:(NSDictionary *)msg) {
    NSLog(@"[BackgroundRunnerModule] Received message: %@", msg);
    [self sendEventWithName:@"toUI" body:msg];
}

RCT_EXPORT_METHOD(sendToBackground:(NSDictionary *)msg) {
    NSLog(@"[BackgroundRunnerModule] Received message: %@", msg);
    [self sendEventWithName:@"toBackground" body:msg];
}

@end

File: apps/mobile/ios/OneKeyWallet/BackgroundRunnerModule.swift
import Foundation
import React
import React_RCTAppDelegate
import ReactAppDependencyProvider

@objc(BackgroundRunnerModule)
class BackgroundRunnerModule: RCTEventEmitter {
  static let shared = BackgroundRunnerModule()
  var reactNativeFactory: RCTReactNativeFactory?
  var reactNativeFactoryDelegate: RCTReactNativeFactoryDelegate?

  private static var isStarted = false
  private var hasListeners = false

  override init() {
    super.init()
  }

  @objc
  static func sharedInstance() -> BackgroundRunnerModule {
    return BackgroundRunnerModule.shared
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  override func supportedEvents() -> [String]! {
    return ["toUI", "toBackground"]
  }

  override func startObserving() {
    hasListeners = true
  }

  override func stopObserving() {
    hasListeners = false
  }

  @objc
  func startBackgroundRunner() {
    if BackgroundRunnerModule.isStarted {
      return
    }
    BackgroundRunnerModule.isStarted = true
    reactNativeFactoryDelegate = BackgroundReactNativeDelegate()
    reactNativeFactoryDelegate!.dependencyProvider = RCTAppDependencyProvider()
    reactNativeFactory = RCTReactNativeFactory(delegate: reactNativeFactoryDelegate!)
    reactNativeFactory!.rootViewFactory.view(withModuleName: BackgroundReactNativeDelegate.BACKGROUND_BUNDLE_NAME, initialProperties: [:])
  }

  @objc(sendToUI:)
  func sendToUI(_ msg: [String: Any]) {
    print("[BackgroundRunnerModule] Received message: \(msg)")
    if hasListeners {
      sendEvent(withName: "toUI", body: msg)
    }
  }

  @objc(sendToBackground:)
  func sendToBackground(_ msg: [String: Any]) {
    print("[BackgroundRunnerModule] Received message: \(msg)")
    if hasListeners {
      sendEvent(withName: "toBackground", body: msg)
    }
  }
}

class BackgroundReactNativeDelegate: RCTDefaultReactNativeFactoryDelegate {
  static let BACKGROUND_BUNDLE_DEBUG_URL = "http://localhost:8082/apps/mobile/background.bundle?platform=ios&dev=true&lazy=false&minify=false&inlineSourceMap=false&modulesOnly=false&runModule=true&excludeSource=true&sourcePaths=url-server&app=so.onekey.wallet&transform.routerRoot=app&transform.engine=hermes&transform.bytecode=1&unstable_transformProfile=hermes-stable"
  static let BACKGROUND_BUNDLE_NAME = "background"
  override func sourceURL(for bridge: RCTBridge) -> URL? {
    bundleURL()
  }


  override func bundleURL() -> URL? {
    #if DEBUG
    return RCTBundleURLProvider.sharedSettings().jsBundleURL(forBundleRoot: BackgroundReactNativeDelegate.BACKGROUND_BUNDLE_NAME)
    #else
      return Bundle.main.url(forResource: BackgroundReactNativeDelegate.BACKGROUND_BUNDLE_NAME, withExtension: "jsbundle")
    #endif
  }
  
  override func createJSRuntimeFactory() -> JSRuntimeFactoryRef {
    return super.createJSRuntimeFactory()
    //return jsrt_create_jsc_factory() // Use JavaScriptCore runtime
  }
}
